PR to fix some problems with the storage

phdussud opened the pull request:
Bunch of fixes to fix the hand over of the file system changes between BP and the host (At least Windows). 
Support for Windows device eject command. 
After all file system object writes, the BP mass storage is ejected and re-inserted to force Windows to notice the changed storage. 
By analyzing the USB MSC traffic between a well known Trenscend SD card reader I found out and implemented the correct sequence of SENSE CODEs to emit during ejection and insertion of a changed storage system.

henrygab commented on nand/nand_ftl_diskio.c:
```diff
@@ -66,7 +68,7 @@ DRESULT diskio_read(BYTE drv, BYTE *buff, LBA_t sector, UINT count)
 
 	if (drv) return STA_NOINIT;		/* Supports only drive 0 */
-
+    mutex_enter_blocking(&diskio_mutex);
```
Can you help me understand what this mutex is protecting against?

I'm trying to see where both CPU cores would enter a state that would cause a conflict.  I'm sure you found something, and want to ensure it's documented, so future code changes don't inadvertently break stuff again through (likely my) ignorance.

phdussud replied:
Dhara isn't meant for concurrent access (not thread safe by design). Core1 is handling USB thus all USBMS operations. Core0 is doing the local file operations. 
It is possible for Core1 to do a write while Core0 is doing a read. The mutex is to protect against this kind of situation.

henrygab replied:
Nice catch!

henrygab commented on msc_disk.c:
```diff
@@ -182,18 +232,15 @@ bool tud_msc_start_stop_cb(uint8_t lun, uint8_t power_condition, bool start, boo
+          // load disk storage
+      }else {
+          host_ejected = true;
+          if (writable)
```
Please use braces with `if`/`else`/`do`/`while` statements.   Not using braces, especially in code that uses macros (e.g., anything in C) can lead to subtle bugs.

henrygab reviewed (changes requested):
Well done getting a trace from a related real device.
Overall, this PR should improve the host's detection of the emulated media state changes.
This would therefore appear to address half the storage corruption problems (the more prominent half) ... **_well done!_**

If you'll humor me, I do have some questions and requested changes.

phdussud commented on msc_disk.c:
```diff
@@ -128,21 +140,64 @@ void tud_msc_inquiry_cb(uint8_t lun, uint8_t vendor_id[8], uint8_t product_id[16
+// This function will be called by core0
+static bool is_ejected()
+{
+    return medium_state == md_state_not_present;
```
No because the host does not know until the SCSCI command REQUEST_SENSE has been issued.

phdussud commented on msc_disk.c:
```diff
@@ -128,21 +140,64 @@ void tud_msc_inquiry_cb(uint8_t lun, uint8_t vendor_id[8], uint8_t product_id[16
+// This function will be called by core0
+static bool is_inserted()
+{
+    return medium_state == md_state_ready;
```
See is_ejected

phdussud commented:
@henrygab. I am very sorry, but I am a little lost now... I did resolve all the PR comments on my github fork, repo, at the same time you merge the PR into main. Now I don't know what to do... What should I do now? Create a new PR? My github repo is in a state that I am not used to dealing with. It is 7 commits behind the official repo and 7 ahead of the official repo. It won't sync.

henrygab commented:
~~You can delete your branch.~~ Oh... you aren't working in a branch.  If you weren't using branches, then the simplest solution may be to just recursively delete your enlistment, and then re-enlist?  There's probably other more complex ways to do this also. 

<details><summary>How I use git</summary><p/>

These simple tips help me use git more effectively.  I share them for free, and you get what you pay for.  Still, may be some hidden wisdom below....

I leave the `main` branch matching the upstream `main` branch (or some point on its chain of commits), and only work in branches.  I do this by always using the `--ff-only` switch when merging the main branch.  This ensures it provides a base commit for my branches that matches a commit on the `upstream/main` branch ... very important for later `rebase` operations.

When starting to do work on an issue, I use the following sequence.  This will first pull the latest changes, then ensure your main branch is up-to-date.  Then it create a new branch for you to make additional changes in (in this example, named `some_name_for_branch`).

```bash
git switch main
git pull --all
git merge --ff-only upstream/main
git switch -c some_name_for_branch
```

Occasionally, to pickup the latest upstream changes:

```bash
git pull --all
git switch main
git merge --ff-only upstream/main
git switch -c some_name_for_branch
```

And then, the magic part.... `rebase`'ing your branch with the latest upstream changes.  Why would I do this?  At least two reasons:
* Get any new upstream fixes that merged into main 
* Reduce the likelihood of merge conflicts when ready to generate a PR

```bash
# identical to the above, gets the latest changes into local main branch first
git switch main
git merge --ff-only upstream/main

# then rebase your working branch so the changes "start at" that
# most recent commit from (your now local) main commit
git switch some_name_for_branch
git rebase main
# resolve conflicts, if any appeared with the recent changes
# (this avoids a PR being blocked due to conflicts)

# NOTE: if you've already published the branch to github,
#            you have to force-push after rebase'ing.  Normally
#            a force-push is to be avoided; Here it's OK since you are
#            the only one modifying the branch.
git push -f
```

</details>

phdussud commented:
Hey, thanks for this tutorial.  I will heed your advice. 
In the meantime, I used the "Create a PR" button and I resolved every conflict (essentially old->new changes in the files I changed for the new PR). Then I deleted and recreated my fork.
